### PR TITLE
UnitTest_Engine: Add methods for filtering unit tests by status and updating existing UnitTest Datasets

### LINKED ISSUE #386: UnitTest_Engine: Add methods for simplify the process of added TestData to an existing UnitTest
#### Description:

When edge cases comes up it is good to capture them in a unit test to make sure they are tested for for any future refactorings. Good to add a method to help simplify this process.

### LINKED ISSUE #403: Add method for getting failing unittests
#### Description:

Usefull to have a method for extracting failing UnitTests from either a collection, a UT dataset, or a file.

This can help with investigation of failing tests if flagged on a PR. Not that hard to do on the GH canvas with a couple of component, but still benefit of a method to streamline the process.

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

 Closes #386 
Closes #403 

 

- Add method(s) for filtering out UnitTests based on their Status. This will help with inspecting tests that have failed by setting status to "Error" or "Warning".
- Update behaviour of "StoreUnitTest" when `replacePreExisting` is false:

previous behaviour for `replacePreExisting == false` when a dataset already exist:
- Return error and exit

new behaviour:
- Pull existing Dataset
- Merge pre-exisitng dataset with new
- Store merged Dataset

To achieve this a few new methods have been added:
- MergeTestDatasets - merges the UTs and SourceInformation of the two datasets. If information is conflicting, the data from the pre-exisiting will be used. For Author, all Authors are added, separated by a new line.
- CullDuplicateTestData - Ensures that the same TestData is not added multiple times to the UT. Think this is an important feature to have for this merging, as one could quite easily accidentally push the same test set twice, and without the culling of duplicate TestData, the list of TestData would grow to contain the same entries twice. If using a toggle rather than a button, and accidentally leaving it to true, while pulling a slider, this could easily mean multiple entries accidentally being added.

 ### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/EnuO6cYoUYFLsUth82UqVsYByNE-LJ2Y_tWuavl9b6dJmw?e=xZEbHE

 ### Changelog

 ### Additional comments

### CODE REVIEW

File: UnitTest_Engine/Compute/MergeTestDataSets.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Merges two Dataset of UnitTests into one Dataset.")]
+        [Input("set1", "First Dataset to merge.")]
+        [Input("set2", "Second Dataset to merge.")]
+        [Output("mergedDataset", "Dataset containing UnitTests and source information from both provided Datasets.")]
+        public static Dataset MergeTestDataSets(Dataset set1, Dataset set2)
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Compute/MergeTestDataSets.cs:44]: Would it make any sense to accept a list of Dataset as an input and merge all datasets in the list into one?

File: UnitTest_Engine/Compute/MergeTestDataSets.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+namespace BH.Engine.UnitTest
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Merges two Dataset of UnitTests into one Dataset.")]
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Compute/MergeTestDataSets.cs:40]:
```suggestion
        [Description("Merges two Datasets of UnitTests into one Dataset.")]
```

File: UnitTest_Engine/Compute/MergeUnitTests.cs [hunk 3]
```diff
@@ -37,10 +37,11 @@ public static partial class Compute
         /**** Public Methods                            ****/
         /***************************************************/
 
+        [PreviousVersion("6.0", "BH.Engine.UnitTest.Compute.MergeUnitTests(System.Collections.Generic.List<BH.oM.Test.UnitTests.UnitTest>)")]
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Compute/MergeUnitTests.cs:40]: Unnecessary - this toolkit isn't in beta so versioning isn't required 😄 but good to see you thinking of it at all times/just in case!

File: UnitTest_Engine/Query/DatasetFromFile.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        [Input("fileName", "Full filepath to the dataset.")]
+        [Output("dataSet", "The DataSet contained in the file.")]
+        public static Dataset DatasetFromFile(string fileName)
+        {
+            if (string.IsNullOrEmpty(fileName))
+            {
+                Base.Compute.RecordError("No filename provided.");
+                return null;
+            }
+            StreamReader sr = new StreamReader(fileName);
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Query/DatasetFromFile.cs:50]:
```suggestion
            }
            
            StreamReader sr = new StreamReader(fileName);
```

File: UnitTest_Engine/Query/FilterByCheckStatus.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Runs and checks all provided UnitTests and provides the ones that has a the provided status.")]
+        [Input("test", "The tests to filter.")]
+        [Input("status", "Status looked for.")]
+        [Input("filterData", "If true, returns new UnitTests objects where the data has been filtered to only contain TestData which returns the provided status. If false, returns full provided UnitTest if the overall status is matching the provided status.")]
+        [MultiOutput(0, "tests", "The filtered tests.")]
+        [MultiOutput(1, "results", "Results matching the filtered tests.")]
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Query/FilterByCheckStatus.cs:53]: Could we do another output for tests which don't match the filter, so we could daisy chain or inspect the other options from the same component?

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Some minor comments/questions but this is good and I'd be happy with it if you disagree with the suggestions - I like the comments in the code too, definitely very helpful 👍 😄

File: UnitTest_Engine/Compute/MergeTestDataSets.cs (hunk 1 above)
Comment (IsakNaslundBh) [reply UnitTest_Engine/Compute/MergeTestDataSets.cs:44]: Can add a method like that as well, just chain calling this method if you like. Still think it is useful to have this one accepting two specific ones as well.

File: UnitTest_Engine/Compute/MergeUnitTests.cs (hunk 3 above)
Comment (IsakNaslundBh) [reply UnitTest_Engine/Compute/MergeUnitTests.cs:40]: Yeah, still like to add it to avoid scripts breaking etc. Know it only has effect for people building this, and then building Versioning_Tk, but feel that is better than nothing!

File: UnitTest_Engine/Query/FilterByCheckStatus.cs (hunk 5 above)
Comment (IsakNaslundBh) [reply UnitTest_Engine/Query/FilterByCheckStatus.cs:53]: hum.. Guess we could, but makes things a bit complex with the toggle input, especially when putting it to true. Might be worth a quick chat

Comment (IsakNaslundBh) [reply UnitTest_Engine/Query/FilterByCheckStatus.cs]: Went ahead and fixed it

File: UnitTest_Engine/Compute/MergeTestDataSets.cs (hunk 1 above)
Comment (FraserGreenroyd) [reply UnitTest_Engine/Compute/MergeTestDataSets.cs:44]: Yeah that's what I'm thinking

Comment (IsakNaslundBh) [reply UnitTest_Engine/Compute/MergeTestDataSets.cs:44]: Done :) https://github.com/BHoM/Test_Toolkit/pull/404/commits/18bc4119b0e63e9f51e82ec0513b144e63a034c2

**Comment by IsakNaslundBh** [comment]
@BHoMBot check unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check null-handling
@BHoMBot check project-compliance

**Review by FraserGreenroyd — DISMISSED** [review]
Code changes make sense, test script looks sensible and know this has been tested against producing unit tests for Geometry_Engine so happy for this to go in to aid the unit test creation work.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check null-handling
@BHoMBot check project-compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot fix project file ref. 8930390188

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check null-handling
@BHoMBot check project-compliance

**Review by FraserGreenroyd — APPROVED** [review]
Only change since last review is compliance so still happy
